File: 序号/232.用栈实现队列.cpp
class MyQueue {
 private:
  stack<int> s1,
      s2;  // s1管插入；
           // 要弹出的时候s1弹出到s2，s2中就是反向的，栈顶就是队列的头。要插入时反向操作
 public:
  MyQueue() {}

  void push(int x) { s1.emplace(x); }

  int pop() {
    if (s2.empty()) {
      while (s1.size()) {
        int i = s1.top();
        s1.pop();
        s2.push(i);
      }
    }
    int i = s2.top();
    s2.pop();
    return i;
  }

  int peek() {
    if (s2.empty()) {
      while (s1.size()) {
        int i = s1.top();
        s1.pop();
        s2.emplace(i);
      }
    }
    return s2.top();
  }

  bool empty() { return s1.empty() && s2.empty(); }
};

File: 序号/232. 用栈实现队列.py
class MyQueue:

    def __init__(self):
        self.s1=[]
        self.s2=[]

    def push(self, x: int) -> None:
        while self.s2:
            self.s1.append(self.s2.pop())
        self.s1.append(x)

    def pop(self) -> int:
        if self.s2:
            return self.s2.pop()
        while len(self.s1)>1:
            self.s2.append(self.s1.pop())
        return self.s1.pop()
        
    def peek(self) -> int:
        if self.s2:
            return self.s2[-1]
        return self.s1[0]

    def empty(self) -> bool:
        return len(self.s1)+len(self.s2)==0



# Your MyQueue object will be instantiated and called as such:
# obj = MyQueue()
# obj.push(x)
# param_2 = obj.pop()
# param_3 = obj.peek()
# param_4 = obj.empty()